extensions/_states/heroku.py

```diff
@@ -0,0 +1,181 @@
+    :returns: three global dicts
+    '''
+
+
+    global diff_heroku_pillar_dict, heroku_diff_dict, pillar_diff_dict
+    app_config_vars_set = set(app_config_vars.items())
+    config_vars_set = set(config_vars.items())
+
+    if app_config_vars_set.symmetric_difference(config_vars_set):
+        diff_heroku_pillar_dict['key_or_value_mismatch'] = True
+        heroku_diff_dict = dict(app_config_vars_set.difference(config_vars_set))
+        pillar_diff_dict = dict(config_vars_set.difference(app_config_vars_set))
+    if len(app_config_vars) - len(config_vars) > 0:
+        diff_heroku_pillar_dict['heroku_not_pillar'] = True
+    if len(app_config_vars) - len(config_vars) < 0:
+        diff_heroku_pillar_dict['pillar_not_heroku'] = True
+
+    log.debug("Completed _diff_heroku_pillar_vars")
+    return diff_heroku_pillar_dict, heroku_diff_dict, pillar_diff_dict
+
+
+def update_app_config_vars(name, config_vars, api_key):
+    '''
+    Update Heroku config vars based on specified pillar config vars.
+    This function will perform the following actions:
+      - Change Heroku config var values for the keys that match one in pillar
+      - Add keys from pillar to Heroku
+      - Leave Heroku keys not find in pillar unchanged
+
+    :param name: The name of the Heroku app
+    :param config_vars: The config_vars specified in pillar
+    :param api_key: The Heroku api_key
+    :returns: Result of executing the state
+    :rtype: dict
+    '''
+
+
+    ret = {'name': name,
+           'comment': '',
+           'result': True,
+           'changes': {}}
+
+    diff_app_config_vars(name, config_vars, api_key)
+    if any(v for v in diff_heroku_pillar_dict.values()):
+        __salt__['heroku.update_app_config_vars'](name, config_vars, api_key)
+        new_app_config_vars = __salt__['heroku.list_app_config_vars'](name, api_key)
+        ret['changes']['new'] = {'heroku': new_app_config_vars}
+    return ret
+
+def override_app_config_vars(name, config_vars, api_key):
+    '''
+    * WARNING * Destructive Function
+    This function will delete all Heroku config var keys and values and
+    replace them with ones specified in pillar
+
+    :param name: The name of the Heroku app
+    :param config_vars: The config_vars specified in pillar
+    :param api_key: The Heroku api_key
+    :returns: Result of executing the state
+    :rtype: dict
+    '''
+
+
+    ret = {'name': name,
+           'comment': 'Heroku config variables have been overwritten',
+           'result': True,
+           'changes': {}}
+
+    try:
+        app_config_vars = __salt__['heroku.list_app_config_vars'](name, api_key)
+        empty_values = dict.fromkeys(app_config_vars)
+        for key, value in empty_values.items():
+            if value is None:
+                values = ''
+                empty_values[key] = value
+
+        log.debug("Created empty_dict")
+        __salt__['heroku.update_app_config_vars'](name, empty_values, api_key)
```

Comment:
Is there any equivalent to the `heroku config:unset` command in the CLI?

Reply:
I think that this would be better as something like
```
app_values.update(config_vars)
__salt__['heroku.update_app_config_vars'](name, app_values, api_key)
```

This way it avoids having the Heroku app restart with a completely empty set of values and will instead do a bulk update and reset of all values at once.

Reply:
As far as I can tell, there is no `unset` equivalent in the [API](https://devcenter.heroku.com/articles/platform-api-reference#config-vars). If I'm not mistaken, running an update with new key values will update those, however it will keep unmatched keys intact.